lib/rust/ensogl/core/src/display/scene.rs

```diff
@@ -1086,6 +1085,7 @@ impl Scene {
                 self.update_shape();
                 self.update_symbols();
                 self.handle_mouse_over_and_out_events();
+                context.shader_compiler.run(time);
```

Comment:
I did it this way on purpose. 

I see multiple problems with the approach proposed in your comment:
1. According to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/WebGL_best_practices#avoid_blocking_api_calls_in_production):
    <img width="768" alt="image" src="https://user-images.githubusercontent.com/1623053/162642647-de8f09f2-e99e-492b-9d81-a989adbed684.png">
    So, it might not be a good idea to evaluate `get*Parameter` functions on the beginning of a frame. 
2. Your `link_check` queue can be big in the compiler. Every `get*Parameter` is costly. You do not know on the beginning of the frame how much time you have left for these operations in this frame, because you don't know how much non-compiler computations are required.
3. This will make the shader compiler pipeline way more complex, while getting results 1 frame faster does not seem like a big win, especially as we are rendering things ~60 frames per second.

Reply:
You are right, the problem here is not `get*Parameter` for `COMPLETION_STATUS_KHR`, but there are several problems that require better description:
- On platforms where KHR extension is not accessible, we would need to use the blocking `get*Parameter` (the compilation can still be performed in parallel by the browser).
- Even if KHR is accessible, we still need to check linking status for each shader, which is costly operation and I think it requires flush + round-trip.
- Even if KHR is accessible, starting new compilations on the beginning of the frame is not a goo idea IMO, because starting the compilation takes some time as well. This actually can be caused by the fact that the shader code in Rust is kept in big Strings that need to be converted to JS and send to the driver, but I'm not sure about that. Anyway, we do not know how many compilations we can spawn at the beginning of the frame.

To sum this up, I see several problems and the solution is not obvious to me. Feel free to help me resolve them and if you see a solution, I'd be happy to implement an improvement here. On the other hand, this will probably make the implementation way more complex and the difference will probably never be spot on in real world cases (as you've written, there are other solutions for the problem of big missed frames).